r1cs/gadgets/std/src/sha256.rs

```diff
@@ -0,0 +1,392 @@
+    )?;
+
+    let h2 = UInt32::addmany(
+        cs.ns(|| "new h2"),
+        &[current_hash_value[2].clone(), c],
+    )?;
+
+    let h3 = UInt32::addmany(
+        cs.ns(|| "new h3"),
+        &[current_hash_value[3].clone(), d],
+    )?;
+
+    let h4 = e.compute(
+        cs.ns(|| "deferred h4 computation"),
+        &[current_hash_value[4].clone()],
+    )?;
+
+    let h5 = UInt32::addmany(
+        cs.ns(|| "new h5"),
+        &[current_hash_value[5].clone(), f],
+    )?;
+
+    let h6 = UInt32::addmany(
+        cs.ns(|| "new h6"),
+        &[current_hash_value[6].clone(), g],
+    )?;
+
+    let h7 = UInt32::addmany(
+        cs.ns(|| "new h7"),
+        &[current_hash_value[7].clone(), h],
+    )?;
+
+    Ok(vec![h0, h1, h2, h3, h4, h5, h6, h7])
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{alloc::AllocGadget, boolean::AllocatedBit, test_constraint_system::TestConstraintSystem};
+    use algebra::fields::bls12_381::Fr;
+ 
+    use rand::{RngCore, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+
+    #[test]
+    fn test_blank_hash() {
+        let iv = get_sha256_iv();
+
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let mut input_bits: Vec<_> = (0..512).map(|_| Boolean::Constant(false)).collect();
+        input_bits[0] = Boolean::Constant(true);
+        let out = sha256_compression_function(&mut cs, &input_bits, &iv).unwrap();
+        let out_bits: Vec<_> = out.into_iter().flat_map(|e| e.into_bits_be()).collect();
+
+        assert!(cs.is_satisfied());
+        assert_eq!(cs.num_constraints(), 0);
+
+        let expected = hex::decode("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855").unwrap();
+
+        let mut out = out_bits.into_iter();
+        for b in expected.iter() {
+            for i in (0..8).rev() {
+                let c = out.next().unwrap().get_value().unwrap();
+
+                assert_eq!(c, (b >> i) & 1u8 == 1u8);
+            }
+        }
+    }
+
+    #[test]
+    fn test_full_block() {
+        let mut rng = XorShiftRng::from_seed([
+            0x59, 0x62, 0xbe, 0x3d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06,
+            0xbc, 0xe5,
+        ]);
+
+        let iv = get_sha256_iv();
+
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let input_bits: Vec<_> = (0..512)
+            .map(|i| {
+                Boolean::from(
+                    AllocatedBit::alloc(
+                        cs.ns(|| format!("input bit {}", i)),
+                        || Ok(rng.next_u32() % 2 != 0),
+                    )
+                    .unwrap(),
+                )
+            })
+            .collect();
+
+        sha256_compression_function(cs.ns(|| "sha256"), &input_bits, &iv).unwrap();
+
+        assert!(cs.is_satisfied());
+        //assert_eq!(cs.num_constraints() - 512, 25840);
+        println!("num_constraints:{}",cs.num_constraints());
+    }
+
+    // #[test]
+    // fn test_against_vectors() {
```

Comment:
Why commented ?

Reply:
It uses sha2 not yet available in our code

Reply:
It's an external dependency, maybe it's easy to import. Would be important to have also the primitive available for testing